### DISCUSSION TITLE
Collect development data for analytics

### DISCUSSION THREAD

**Discussion by sergfeldman** [ubiquity/ubiquibot#154] [Ideas]
This is the draft of the issue ready for review.
Proposals of the additional data are highly appreciated.

### Background 
In order to esablish development analytics and reporting it is required to collect data available on GitHub.

The discussion 
https://github.com/orgs/ubiquity/discussions/15

### To do 

Establish the collecting the defined below data from the GitHub to a place where it is possible to perform analytics and reporting.

The place should be defined during the implementation - a database, GitHub itself, solutions like ELK, or something else.

The metrics for the reporting can be found in the discussion ^

**Data structure**
- ID
- Entity
- Action title
- Action date and time 
- GitHub handle that performed the action

**Entities**
- Issue
- Code
- CI/CD

**Action titles**

Issue actions 
- change an assignee
- change status (to do, in progress, review, close as completed, close as not planned, etc)
- link PR (and create a branch with this issue)
- post a comment
- add and change Time labels (to analyze on-time delivery)

Code actions
- open PR
- add a commit
- get a first reaction to the PR (comment/approve)
- request changes (need to check if this is possible to track)
- merge PR
- close PR
- revert PR (need to check if this is possible to track)

CI/CD actions 
- status of the regular Main branch check of repo stability (perform the regular check every day, need to clarify the way to check - for example, automatic build with pased autotests or somehow else)
- performed release (build and deploy to the Production environment)

**Comment by 0x4007** [comment]
I encourage you to prototype this at a smaller scale by semi manually aggregating those data points either by hand or ideally with some third party tool/their website (think of whatthediff.ai having the demo where you can simply paste in the link to the repo and it will spit out analytics)

**Comment by sergfeldman** [comment]
GitHub analytics 
GitHub repo analysis tools
Git analytics tools
GihHub repo development analytic 
GitHub repo analyyic for Engineering managers 
github repo metrics mining tools

https://documentation.wazuh.com/current/github/monitoring-github-activity.html
https://www.mezmo.com/blog/monitoring-github-activity-with-logdna
http://janvitek.org/events/NEU/6050/github_download.html
https://keen.io/docs/integrations/github/
https://github.com/erikbern/git-of-theseus
https://github.com/feedzai/repo-analyzer
https://github.com/reposense/RepoSense
https://console.cloud.google.com/marketplace/details/github/github-repos?pli=1
https://docs.github.com/en/enterprise-server@3.5/admin/monitoring-activity-in-your-enterprise/exploring-user-activity
https://docs.github.com/en/rest/metrics/statistics?apiVersion=2022-11-28
https://docs.github.com/en/enterprise-cloud@latest/admin/monitoring-activity-in-your-enterprise/reviewing-audit-logs-for-your-enterprise/exporting-audit-log-activity-for-your-enterprise

https://www.cdata.com/kb/tech/github-odbc-excel-query.rst

https://www.gitclear.com/pricing

https://github.com/engineering-management/awesome-engineering-management#kpi-key-performance-indicators

**Reply by 0x4007** [reply]
I'm not sure how capable you are at setting up code but with the repos you can import them into replit to run them quickly. I haven't used replit but I read about it and I guess its for this sort of usecase.

https://replit.com/@pavlovcik/repo-analyzer

**Comment by sergfeldman** [comment]
### Tools for the implementation as a product 

Open Source BI tools

https://github.com/getredash/redash
https://redash.io/product/

https://www.metabase.com/pricing/

**Comment by sergfeldman** [comment]
### Tools and reading for the implementation for Ubiquity

**ELK**
https://github.com/grafana/github-to-es
https://grafana.com/grafana/dashboards/2250-github-repo-trends-comments/

**Systems**

Ready to use Open Source development analytics CHAOSS - Grimoire lab
https://chaoss.community/
https://chaoss.community/kb-metrics-and-metrics-models/
https://chaoss.community/software/
https://chaoss.github.io/grimoirelab/
https://github.com/chaoss
https://github.com/chaoss/grimoirelab
https://github.com/chaoss/augur

Require permissions from GitHub repo admin 
https://compass.gitee.com/docs/quick-start/
https://github.com/oss-compass

Require permissions from GitHub repo admin
https://cauldron.io/
https://cauldron.io/projects/new
https://cauldron.io/pricing/

**Libraries**

https://pydriller.readthedocs.io/en/latest/
https://github.com/ishepard/pydriller
https://accessibleai.dev/post/extracting-git-data-pydriller/
https://github.com/gitpython-developers/GitPython

https://github.com/ejwa/gitinspector

https://github.com/akashraj9828/gitstats

https://github.com/mauricioaniche/repodriller

**Addional tools**
https://www.mergestat.com/pricing
https://github.com/mergestat

https://sourcecred.io/

https://gitcompare.com/

https://www.aboutcode.org/

https://github.com/gotec/git2net

https://github.com/Netflix/osstracker

**Paid or closed sourced solutions**
https://waydev.co/pricing/
Cool features include
https://waydev.co/features/pr-workflow-analytics/ 
https://snoot.io/


**Reading**
Data-driven Engineering management, Git Analytics for Engineering Leads
https://www.hatica.io/blog/engineering-teams-and-git-analytics/
https://www.hatica.io/blog/engineering-analytics-for-well-being/
https://nimblehq.co/compass/development/git-analytics/engineering-lead/

**Next research steps**
Lists of tools and approach
https://www.revelo.com/blog/best-software-development-analytics-tools
https://livablesoftware.com/tools-mine-analyze-github-git-software-data/
https://levelup.gitconnected.com/how-to-mine-github-data-in-2022-e9c70b3f61d3

The example of how development analytics is used for reports ie investment reports
https://bitergia.com/resources/#reports
